crates/runtime/src/instance/allocator/pooling/unix.rs

```diff
@@ -1,26 +1,45 @@
 use anyhow::{Context, Result};
+use rustix::mm::{mprotect, MprotectFlags};
 
 fn decommit(addr: *mut u8, len: usize, protect: bool) -> Result<()> {
     if len == 0 {
         return Ok(());
     }
 
-    // By creating a new mapping at the same location, this will discard the
-    // mapping for the pages in the given range.
-    // The new mapping will be to the CoW zero page, so this effectively
-    // zeroes the pages.
     unsafe {
-        rustix::mm::mmap_anonymous(
-            addr as _,
-            len,
-            if protect {
-                rustix::mm::ProtFlags::empty()
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                use rustix::mm::{madvise, Advice};
+
+                if protect {
+                    mprotect(addr.cast(), len, MprotectFlags::empty())
+                        .context("failed to protect memory pages")?;
+                }
+
+                // On Linux, this is enough to cause the kernel to initialize
+                // the pages to 0 on next access
+                madvise(addr as _, len, Advice::LinuxDontNeed)
+                    .context("madvise failed to decommit: {}")?;
             } else {
-                rustix::mm::ProtFlags::READ | rustix::mm::ProtFlags::WRITE
-            },
-            rustix::mm::MapFlags::PRIVATE | rustix::mm::MapFlags::FIXED,
-        )
-        .context("mmap failed to remap pages: {}")?;
+                use rustix::mm::{mprotect, MprotectFlags};
+
+                // By creating a new mapping at the same location, this will
+                // discard the mapping for the pages in the given range.
+                // The new mapping will be to the CoW zero page, so this
+                // effectively zeroes the pages.
+                mmap_anonymous(
+                    addr as _,
+                    len,
+                    if protect {
+                        ProtFlags::empty()
+                    } else {
+                        ProtFlags::READ | rustix::mm::ProtFlags::WRITE
+                    },
+                    MapFlags::PRIVATE | rustix::mm::MapFlags::FIXED,
```

Comment:
These extra `rustix::mm::*` qualifiers are redundant because `ProtFlags` and `MapFlags` are already in scope.